### DISCUSSION TITLE
Running QtApp on native c++ engine + rust backend w/o QML-engine

### DISCUSSION THREAD

**Discussion by Kuraga13** [KDAB/cxx-qt#647] [Q&A]
Hello,
I am trying to do something similar to the example [demo_thread](https://github.com/KDAB/cxx-qt/tree/main/examples/demo_threading).
Idea is backend in rust(finally in a separate thread as in the example), and frontend-application in Qt.I mean with a ready-made working Qt application.
Now I'm trying to collect at least something minimal demo.
```
use core::pin::Pin;

#[cxx_qt::bridge(cxx_file_stem = "backend_interface", namespace = "cxx_qt::backend_interface")]
pub mod qobject {

    extern "RustQt" {
        #[qobject]
        type BackendInterface = super::BackendInterfaceRust;
    }

    unsafe extern "RustQt" {
        #[qinvokable]
        fn test_backend(self: Pin<&mut BackendInterface>);
    }
}

#[derive(Default)]
pub struct BackendInterfaceRust;

impl qobject::BackendInterface {
    pub fn test_backend(self: Pin<&mut Self>) {

        println!("Hi from Rust! test_backend" );

    }
}
```
```
#include "cxx-qt-gen/backend_interface.cxxqt.h"
MainWindow::MainWindow(QWidget *parent, Qt::WindowFlags flags):
QMainWindow(parent, flags)
{ 
         .......
	cxx_qt::backend_interface::BackendInterface test;
        test.testBackend();
}
```
By the output of the compiler, does it necessarily need a qml engine to work?
```
[build] C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/13.2.0/../../../../i686-w64-mingw32/bin/ld.exe: libtest_qt_rust.a(qqmlapplicationengine.o):qqmlapplicationengine.cpp:(.text$_ZN4rust9cxxqtlib124qqmlapplicationengineNewEv+0x24): undefined reference to 
```
1. Is it supported, is it possible?
2. If yes, can you provide a minimal example of binding or suggest how to do it. thanks in advance

**Comment by Be-ing** [comment]
If you don't want to link to QtQml, disable the `qt_qml` feature of the cxx-qt-lib crate. It's enabled by default, so you'll need `default-features = false` in Cargo.toml.

Regardless, that link error is curious. It looks like the error got cut off? Did you copy and paste the whole error?

**Comment by ahayzen-kdab** [comment]
Note that in addition to demo_threading there is a short example in https://github.com/KDAB/cxx-qt/blob/main/examples/qml_features/rust/src/threading.rs of how to do threading.

And is this a CMake build or a Cargo build you are doing ? this should be possible it just sounds like either you have QML features enabled or we have a bug somewhere that is trying to bind to the QML engine.

**Comment by Kuraga13** [comment]
@Be-ing & @ahayzen-kdab  thanks for the quick replies. Yes, I didn't have `default-features = false` in `cxx-qt-lib`, now all is work, thanks. 

To build my rather old project from Qt5, I had to take into account some little things. For example setting Q_SLOTS instead of slots.
Now it remains to connect qt_thread, I will keep this discussion open - if something works out, I will create a small-repository with an example.

**Reply by ahayzen-kdab** [reply]
For the `Q_SLOTS` vs `slots`, `slots` can still work in C++ code if you haven't set `QT_NO_KEYWORDS` in CMake, we just set that to ensure we don't accidentally use `signals` or `slots` in our generated C++ code which can collide with other frameworks / libraries. Or is this not what you are referring too ?

**Reply by Kuraga13** [reply]
Yes, that's it, thanks.
